File: internal/prodgate/e2e_report.go
package prodgate

import (
	"encoding/json"
	"os"

	"reporting-db-migrations/internal/types"
)

// E2EScenarioReport is the wire format for Go↔Rust scenario e2e (behavior + phase timings).
type E2EScenarioReport struct {
	Scenario     string         `json:"scenario"`
	SetupSteps   []string       `json:"setup_steps,omitempty"`
	Timings      PhaseTimings   `json:"timings"`
	Io           DbIoProfile    `json:"io"`
	Snapshot     PlanSnapshot   `json:"snapshot"`
	ActionCounts map[string]int `json:"action_counts"`
}

// E2EApplyReport is the wire format for migrate/apply outcome e2e.
type E2EApplyReport struct {
	Scenario        string   `json:"scenario"`
	SetupSteps      []string `json:"setup_steps,omitempty"`
	Applied         int      `json:"applied"`
	Failed          int      `json:"failed"`
	Skipped         int      `json:"skipped"`
	Errors          []string `json:"errors,omitempty"`
	AuditObjectRows int      `json:"audit_object_rows"`
}

// E2EGateReport is the wire format for prod gate e2e (Go↔Rust).
type E2EGateReport struct {
	Scenario   string       `json:"scenario"`
	SetupSteps []string     `json:"setup_steps,omitempty"`
	GateGo     bool         `json:"gate_go"`
	Messages   []string     `json:"messages,omitempty"`
	Snapshot   PlanSnapshot `json:"snapshot"`
	Gate       GateResult   `json:"gate"`
}

// E2EBlockedReport is the wire format for blocked migrate + scaffold e2e.
type E2EBlockedReport struct {
	Scenario      string   `json:"scenario"`
	SetupSteps    []string `json:"setup_steps,omitempty"`
	ExitCode      int      `json:"exit_code"`
	Blocked       bool     `json:"blocked"`
	Blockers      []string `json:"blockers,omitempty"`
	ScaffoldPaths []string `json:"scaffold_paths,omitempty"`
}

// ActionCountsFromPlan tallies planned_action values (scenario behavior summary).
func ActionCountsFromPlan(plan *types.MigrationPlan) map[string]int {
	out := map[string]int{}
	if plan == nil {
		return out
	}
	for _, obj := range plan.Objects {
		out[obj.PlannedAction]++
	}
	return out
}

// BuildE2EScenarioReport builds a comparable report for pipeline-level e2e.
func BuildE2EScenarioReport(scenario string, plan *types.MigrationPlan, timings PhaseTimings, io DbIoProfile) E2EScenarioReport {
	return E2EScenarioReport{
		Scenario:     scenario,
		Timings:      timings,
		Io:           io,
		Snapshot:     SnapshotFromPlan(plan),
		ActionCounts: ActionCountsFromPlan(plan),
	}
}

// WriteE2EReportFile writes report JSON (mode 0644).
func WriteE2EReportFile(path string, rep E2EScenarioReport) error {
	data, err := json.MarshalIndent(rep, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// BuildE2EApplyReport builds apply outcome report for e2e.
func BuildE2EApplyReport(scenario string, applied, failed, skipped, auditRows int, errors []string) E2EApplyReport {
	return E2EApplyReport{
		Scenario:        scenario,
		Applied:         applied,
		Failed:          failed,
		Skipped:         skipped,
		Errors:          errors,
		AuditObjectRows: auditRows,
	}
}

// BuildE2EGateReport builds gate e2e report from Evaluate result.
func BuildE2EGateReport(scenario string, snap PlanSnapshot, result GateResult) E2EGateReport {
	return E2EGateReport{
		Scenario: scenario,
		GateGo:   result.Go,
		Messages: append([]string(nil), result.Messages...),
		Snapshot: snap,
		Gate:     result,
	}
}

// WriteE2EApplyReportFile writes E2EApplyReport JSON.
func WriteE2EApplyReportFile(path string, rep E2EApplyReport) error {
	data, err := json.MarshalIndent(rep, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// ReadE2EApplyReportFile loads E2EApplyReport from path.
func ReadE2EApplyReportFile(path string) (E2EApplyReport, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return E2EApplyReport{}, err
	}
	var rep E2EApplyReport
	if err := json.Unmarshal(data, &rep); err != nil {
		return E2EApplyReport{}, err
	}
	return rep, nil
}

// WriteE2EGateReportFile writes E2EGateReport JSON.
func WriteE2EGateReportFile(path string, rep E2EGateReport) error {
	data, err := json.MarshalIndent(rep, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// ReadE2EGateReportFile loads E2EGateReport from path.
func ReadE2EGateReportFile(path string) (E2EGateReport, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return E2EGateReport{}, err
	}
	var rep E2EGateReport
	if err := json.Unmarshal(data, &rep); err != nil {
		return E2EGateReport{}, err
	}
	if rep.Snapshot.Objects == nil {
		rep.Snapshot.Objects = map[string]ObjectEntry{}
	}
	return rep, nil
}

// WriteE2EBlockedReportFile writes E2EBlockedReport JSON.
func WriteE2EBlockedReportFile(path string, rep E2EBlockedReport) error {
	data, err := json.MarshalIndent(rep, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// ReadE2EBlockedReportFile loads E2EBlockedReport from path.
func ReadE2EBlockedReportFile(path string) (E2EBlockedReport, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return E2EBlockedReport{}, err
	}
	var rep E2EBlockedReport
	if err := json.Unmarshal(data, &rep); err != nil {
		return E2EBlockedReport{}, err
	}
	return rep, nil
}

// ReadE2EReportFile loads E2EScenarioReport from path.
func ReadE2EReportFile(path string) (E2EScenarioReport, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return E2EScenarioReport{}, err
	}
	var rep E2EScenarioReport
	if err := json.Unmarshal(data, &rep); err != nil {
		return E2EScenarioReport{}, err
	}
	if rep.ActionCounts == nil {
		rep.ActionCounts = map[string]int{}
	}
	if rep.Snapshot.Objects == nil {
		rep.Snapshot.Objects = map[string]ObjectEntry{}
	}
	return rep, nil
}

File: rust/crates/core/src/gate/e2e_report.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::driver::IoProfile;
use crate::export::MigrationPlan;
use crate::timings::PhaseTimings;

use super::snapshot::PlanSnapshot;

#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct E2EScenarioReport {
    pub scenario: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub setup_steps: Vec<String>,
    pub timings: PhaseTimings,
    pub io: IoProfile,
    pub snapshot: PlanSnapshot,
    pub action_counts: HashMap<String, i32>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct E2EApplyReport {
    pub scenario: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub setup_steps: Vec<String>,
    pub applied: i32,
    pub failed: i32,
    pub skipped: i32,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub errors: Vec<String>,
    pub audit_object_rows: i32,
}

#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct E2EGateReport {
    pub scenario: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub setup_steps: Vec<String>,
    pub gate_go: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub messages: Vec<String>,
    pub snapshot: PlanSnapshot,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct E2EBlockedReport {
    pub scenario: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub setup_steps: Vec<String>,
    pub exit_code: i32,
    pub blocked: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub blockers: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub scaffold_paths: Vec<String>,
}

pub fn action_counts_from_plan(plan: &MigrationPlan) -> HashMap<String, i32> {
    let mut out = HashMap::new();
    if !plan.rows.is_empty() {
        for row in &plan.rows {
            let action = serde_json::to_string(&row.planned_action())
                .unwrap_or_default()
                .trim_matches('"')
                .to_string();
            *out.entry(action).or_insert(0) += 1;
        }
        return out;
    }
    for obj in &plan.objects {
        let action = serde_json::to_string(&obj.planned_action)
            .unwrap_or_default()
            .trim_matches('"')
            .to_string();
        *out.entry(action).or_insert(0) += 1;
    }
    out
}

pub fn build_e2e_report(
    scenario: &str,
    plan: &MigrationPlan,
    timings: &PhaseTimings,
    io: &IoProfile,
) -> E2EScenarioReport {
    E2EScenarioReport {
        scenario: scenario.into(),
        setup_steps: Vec::new(),
        timings: timings.clone(),
        io: io.clone(),
        snapshot: PlanSnapshot::from_plan(plan),
        action_counts: action_counts_from_plan(plan),
    }
}

pub fn read_e2e_report_json(data: &str) -> Result<E2EScenarioReport, serde_json::Error> {
    serde_json::from_str(data)
}

pub fn read_e2e_apply_json(data: &str) -> Result<E2EApplyReport, serde_json::Error> {
    serde_json::from_str(data)
}

pub fn read_e2e_gate_json(data: &str) -> Result<E2EGateReport, serde_json::Error> {
    serde_json::from_str(data)
}

pub fn read_e2e_blocked_json(data: &str) -> Result<E2EBlockedReport, serde_json::Error> {
    serde_json::from_str(data)
}

pub fn write_e2e_report_file(
    path: &std::path::Path,
    rep: &E2EScenarioReport,
) -> std::io::Result<()> {
    write_json_file(path, rep)
}

pub fn write_e2e_apply_file(path: &std::path::Path, rep: &E2EApplyReport) -> std::io::Result<()> {
    write_json_file(path, rep)
}

pub fn write_e2e_gate_file(path: &std::path::Path, rep: &E2EGateReport) -> std::io::Result<()> {
    write_json_file(path, rep)
}

pub fn write_e2e_blocked_file(
    path: &std::path::Path,
    rep: &E2EBlockedReport,
) -> std::io::Result<()> {
    write_json_file(path, rep)
}

fn write_json_file<T: Serialize>(path: &std::path::Path, rep: &T) -> std::io::Result<()> {
    let data = serde_json::to_string_pretty(rep).map_err(std::io::Error::other)?;
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(path, format!("{data}\n"))
}

/// Compare Go reference report vs Rust run: behavior (actions + snapshot) + phase timings.
pub fn compare_e2e_reports(go: &E2EScenarioReport, rust: &E2EScenarioReport) -> Vec<String> {
    let mut msgs = Vec::new();
    if go.scenario != rust.scenario {
        msgs.push(format!(
            "scenario name: go={} rust={}",
            go.scenario, rust.scenario
        ));
    }
    if go.action_counts != rust.action_counts {
        msgs.push(format!(
            "action_counts: go={:?} rust={:?}",
            go.action_counts, rust.action_counts
        ));
    }
    msgs.extend(super::compare::parity_messages(
        &go.snapshot,
        &rust.snapshot,
    ));
    msgs.extend(compare_timings(&go.timings, &rust.timings));
    msgs
}

pub fn compare_e2e_apply_reports(go: &E2EApplyReport, rust: &E2EApplyReport) -> Vec<String> {
    let mut msgs = Vec::new();
    if go.scenario != rust.scenario {
        msgs.push(format!(
            "scenario: go={} rust={}",
            go.scenario, rust.scenario
        ));
    }
    if go.applied != rust.applied {
        msgs.push(format!("applied: go={} rust={}", go.applied, rust.applied));
    }
    if go.failed != rust.failed {
        msgs.push(format!("failed: go={} rust={}", go.failed, rust.failed));
    }
    if go.skipped != rust.skipped {
        msgs.push(format!("skipped: go={} rust={}", go.skipped, rust.skipped));
    }
    if go.audit_object_rows != rust.audit_object_rows {
        msgs.push(format!(
            "audit_object_rows: go={} rust={}",
            go.audit_object_rows, rust.audit_object_rows
        ));
    }
    msgs
}

pub fn compare_e2e_gate_reports(go: &E2EGateReport, rust: &E2EGateReport) -> Vec<String> {
    let mut msgs = Vec::new();
    if go.scenario != rust.scenario {
        msgs.push(format!(
            "scenario: go={} rust={}",
            go.scenario, rust.scenario
        ));
    }
    if go.gate_go != rust.gate_go {
        msgs.push(format!("gate_go: go={} rust={}", go.gate_go, rust.gate_go));
    }
    msgs.extend(super::compare::parity_messages(
        &go.snapshot,
        &rust.snapshot,
    ));
    msgs
}

pub fn compare_e2e_blocked_reports(go: &E2EBlockedReport, rust: &E2EBlockedReport) -> Vec<String> {
    let mut msgs = Vec::new();
    if go.scenario != rust.scenario {
        msgs.push(format!(
            "scenario: go={} rust={}",
            go.scenario, rust.scenario
        ));
    }
    if go.exit_code != rust.exit_code {
        msgs.push(format!(
            "exit_code: go={} rust={}",
            go.exit_code, rust.exit_code
        ));
    }
    if go.blocked != rust.blocked {
        msgs.push(format!("blocked: go={} rust={}", go.blocked, rust.blocked));
    }
    if go.blocked && rust.scaffold_paths.is_empty() && !go.scaffold_paths.is_empty() {
        msgs.push("rust: expected scaffold_paths after blocked migrate".into());
    }
    if !go.scaffold_paths.is_empty() && rust.scaffold_paths.is_empty() {
        msgs.push(format!(
            "scaffold_paths: go={} rust=0",
            go.scaffold_paths.len()
        ));
    }
    msgs
}

fn compare_timings(go: &PhaseTimings, rust: &PhaseTimings) -> Vec<String> {
    let factor = std::env::var("RMIG_E2E_TIMING_FACTOR")
        .ok()
        .and_then(|s| s.parse::<f64>().ok())
        .unwrap_or(3.0);
    let slack_ms: i64 = std::env::var("RMIG_E2E_TIMING_SLACK_MS")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100);

    let phases: &[(&str, i64, i64)] = &[
        (
            "parallel_wall_ms",
            go.parallel_wall_ms,
            rust.parallel_wall_ms,
        ),
        ("diff_ms", go.diff_ms, rust.diff_ms),
        ("plan_wall_ms", go.plan_wall_ms, rust.plan_wall_ms),
    ];

    let mut msgs = Vec::new();
    for (name, go_ms, rust_ms) in phases {
        if *go_ms == 0 && *rust_ms == 0 {
            continue;
        }
        let ceiling = ((*go_ms as f64) * factor).ceil() as i64 + slack_ms;
        if *rust_ms > ceiling {
            msgs.push(format!(
                "{name}: rust={rust_ms}ms > go={go_ms}ms ceiling={ceiling}ms (factor={factor}, slack={slack_ms}ms)"
            ));
        }
    }
    msgs
}
